certora/specs/HashConsensus.spec

```diff
@@ -0,0 +1,742 @@
+}
+
+
+// 24. updateInitialEpoch() - updates correctly the initialEpoch as returned by getFrameConfig()
+//                          - you cannot update the initialEpoch to be one that already it arrived
+// Status: Pass
+// https://prover.certora.com/output/80942/3b2b054cad054559b81de132dd676e0d/?anonymousKey=b5142a9cc196ffc7338c450c4ace6b200fc8d12f
+rule updateInitialEpochCorrectness() {
+    env e1; env e2;
+
+    require e1.block.timestamp < e2.block.timestamp;        // time moves forward only
+
+    // Before update
+    uint256 initialEpoch1; uint256 epochsPerFrame1; uint256 fastLaneLengthSlots1;
+    initialEpoch1, epochsPerFrame1, fastLaneLengthSlots1 = getFrameConfig(e1);
+
+    uint256 initialEpochUpdate;
+    updateInitialEpoch(e2, initialEpochUpdate);
+
+    // After update
+    uint256 initialEpoch2; uint256 epochsPerFrame2; uint256 fastLaneLengthSlots2;
+    initialEpoch2, epochsPerFrame2, fastLaneLengthSlots2 = getFrameConfig(e2);
+
+    assert initialEpochUpdate == initialEpoch2;  // initialEpoch was updated correctly
+    assert (computeEpochAtTimestamp(e2, e2.block.timestamp) < initialEpoch1);  // initialEpoch1 still has not arrived: verifies error InitialEpochAlreadyArrived
+}
+
+
+// 25. initialEpochSanity rule - once the system is initialized its initialEpoch setting is always sane
+// Status: Pass
+// https://prover.certora.com/output/80942/0dec7449b4c049a4b6cca667f094aa91/?anonymousKey=fd75c248d93e5af85a046ca5b1c42723eea43d60
+rule initialEpochSanity(method f) 
+    filtered { f -> !f.isView && 
+                    f.selector != updateInitialEpoch(uint256).selector } // this should revert once initialized (see rule 26)
+{
+    env e; calldataarg args; env e2;
+    require e.block.timestamp > saneTimeConfig();       // time moves forward (saneTimeConfig is for correct initializing)
+
+    f(e,args);                                          // can call any state changing method
+
+    require e2.block.timestamp >= e.block.timestamp;    // after the call above, verify that the initialEpoch is sane
+
+    uint256 slotsPerEpoch; uint256 secondsPerSlot; uint256 genesisTime;
+    slotsPerEpoch, secondsPerSlot, genesisTime = getChainConfig(e2);
+    
+    uint256 initialEpoch; uint256 epochsPerFrame; uint256 fastLaneLengthSlots;
+    initialEpoch, epochsPerFrame, fastLaneLengthSlots = getFrameConfig(e2);
+
+    assert initialEpoch <= ((e2.block.timestamp - genesisTime) / (secondsPerSlot * slotsPerEpoch)); // sane configuration of initialEpoch
+}
+
+
+// 26. updateInitialEpoch should revert once the system is initialized correctly and the initialEpoch passed
+// Status: Pass
+// https://prover.certora.com/output/80942/5c86836015fc4c9baa0cb035e635975a/?anonymousKey=86dce987e054b75ffa9e119ff360e6a47d607156
+rule updateInitialEpochRevertsCorrectly() {
+    env e; calldataarg args;
+    require e.block.timestamp > saneTimeConfig();       // time moves forward (saneTimeConfig is for correct initializing)
```

Comment:
Is there a lighter restriction that will verify this rule?
It's not really necessary to assume everything here, right?